src/libraries/System.Runtime.Numerics/src/System/Numerics/BigNumber.cs

```diff
@@ -494,35 +494,263 @@ private static bool HexNumberToBigInteger(ref BigNumberBuffer number, out BigInt
+                            _partialDigitCount++;
+                            _totalDigitCount++;
+
+                            // Update the buffer when enough partial digits have been accumulated.
+                            if (_partialDigitCount == MaxPartialDigits)
+                            {
+                                MultiplyAdd(ref currentBuffer, TenPowMaxPartial, _partialValue);
+                                _partialValue = 0;
+                                _partialDigitCount = 0;
+                            }
+                        }
+
+                        // Check for nonzero digits after the decimal point.
+                        if (!endReached)
+                        {
+                            ReadOnlySpan<char> fracDigitsSpan = chunkDigits.Slice(intDigitsSpan.Length);
+                            for (int i = 0; i < fracDigitsSpan.Length; i++)
+                            {
+                                char digitChar = fracDigitsSpan[i];
+                                if (digitChar == '\0')
+                                {
+                                    break;
+                                }
+                                if (digitChar != '0')
+                                {
+                                    return false;
+                                }
+                            }
+                        }
+
+                        partialValue = _partialValue;
+                        partialDigitCount = _partialDigitCount;
+                        totalDigitCount = _totalDigitCount;
+
+                        return true;
+                    }
+                }
+                else
                 {
-                    if (!ProcessChunk(digitsChunk.Span, ref currentBuffer))
+                    if (numberScale < 0)
                     {
                         result = default;
                         return false;
                     }
-                }
+                    totalDigitCount = Math.Min(number.digits.Length - 1, numberScale);
+                    int bufferSize = (totalDigitCount + MaxPartialDigits - 1) / MaxPartialDigits;
+
+                    Span<uint> buffer = new uint[bufferSize];
+
+                    // Separate every MaxPartialDigits digits and store them in the buffer.
+                    // Buffers are treated as little-endian. That means, the array { 234567890, 1 }
+                    // represents the number 1234567890.
+                    int bufferIndex = buffer.Length - 1;
+                    uint currentBlock = 0;
+                    int shiftUntil = (totalDigitCount - 1) % MaxPartialDigits;
+                    int remainingIntDigitCount = totalDigitCount;
+                    foreach (ReadOnlyMemory<char> digitsChunk in number.digits.GetChunks())
+                    {
+                        var digitsChunkSpan = digitsChunk.Span;
```

Comment:
Style: Inappropriate use of `var` (the short, imprecise form: the type name needs to appear on the line).

```suggestion
                        ReadOnlySpan<char> digitsChunkSpan = digitsChunk.Span;
```